src/communication/websocket/session.cpp

```diff
@@ -86,14 +106,68 @@ void Session::DoRead() {
       }));
 }
 
-void Session::OnRead(boost::beast::error_code ec, size_t /*bytes_transferred*/) {
+void Session::DoClose() {
+  ws_.async_close(boost::beast::websocket::close_code::normal,
+                  boost::asio::bind_executor(strand_, [shared_this = shared_from_this()](boost::beast::error_code ec) {
+                    shared_this->OnClose(ec);
+                  }));
+}
+
+void Session::OnClose(boost::beast::error_code ec) {
+  if (ec) {
+    return LogError(ec, "close");
+  }
+  connected_.store(false, std::memory_order_relaxed);
+}
+
+bool Session::Authenticate(const nlohmann::json &creds) {
+  return auth_.Authenticate(creds.at("username").get<std::string>(), creds.at("password").get<std::string>());
+}
+
+bool Session::Authorize(const nlohmann::json &creds) {
+  return auth_.DoesUserHasPermission(creds.at("username").get<std::string>(), auth::Permission::WEBSOCKET);
+}
```

Comment:
This can be a single method `Authenticate` which simply checks if the user exists and, if it's enterprise build, checks if he has the correct permissions.

Reply:
I made it like this since we want to form different response messages depending on if the Authentication or Authorization fails. Do you think it better to pass ref to reponse and let them change the message?

Reply:
You can use `utils::BasicResult<std::string>` for that (returns an error message if something went wrong).